File: 0525-Contiguous Array/0525-Contiguous Array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int, int> table;
        table[0] = -1;
        int maxLen = 0;
        int sum = 0;
        for (int i = 0; i < nums.size(); ++i)
        {
            if (nums[i])
            {
                ++sum;
            }
            else
            {
                --sum;
            }
            if (table.find(sum) != table.end())
            {
                maxLen = max(maxLen, i - table[sum]);
            }
            else
            {
                table[sum] = i;
            }
        }
        
        return maxLen;
    }
};

File: 0501-0600/0525-Contiguous Array/0525-Contiguous Array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        count = maxLen = 0
        table = {0:-1}
        for i, num in enumerate(nums):
            count += 1 if num else -1
            if count in table:
                maxLen = max(maxLen, i - table[count])
            else:
                table[count] = i
        return maxLen
